torch/csrc/distributed/c10d/frontend.cpp

```diff
@@ -0,0 +1,633 @@
+    return nullptr;
+  }
+
+  auto work = group->allgather(output_tensor_lists, input_tensor_list);
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::allGather(
+    std::vector<at::Tensor>& tensor_list,
+    at::Tensor tensor,
+    const std::shared_ptr<ProcessGroup>& group,
+    bool async_op) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  std::vector<std::vector<at::Tensor>> output_tensor_lists = {std::move(tensor_list)};
+  std::vector<at::Tensor> input_tensor_list = {std::move(tensor)};
+  auto work = group->allgather(output_tensor_lists, input_tensor_list);
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::allGatherCoalesced(
+    std::vector<std::vector<at::Tensor>>& output_tensor_lists,
+    std::vector<at::Tensor>& input_tensor_list,
+    const std::shared_ptr<ProcessGroup>& group,
+    bool async_op) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  auto work =
+      group->allgather_coalesced(output_tensor_lists, input_tensor_list);
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::gather(
+    at::Tensor tensor,
+    const c10::optional<std::vector<at::Tensor>>& gather_list,
+    const std::shared_ptr<ProcessGroup>& group,
+    int64_t dst,
+    bool async_op) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  auto my_rank = group->getRank();
+
+  std::vector<std::vector<at::Tensor>> output_tensors;
+
+  if (dst == my_rank) {
+    TORCH_CHECK(
+        gather_list.has_value(),
+        "Argument ``gather_list`` must be specified on destination rank");
+    output_tensors.push_back(gather_list.value());
+  } else {
+    TORCH_CHECK(
+        !gather_list.has_value(),
+        "Argument ``gather_list`` must NOT be specified on non-destination ranks.");
+  }
+
+  std::vector<at::Tensor> input_tensors = {std::move(tensor)};
+
+  GatherOptions opts;
+  opts.rootRank = dst;
+
+  std::shared_ptr<ProcessGroup::Work> work;
+  if (group == default_pg_) {
+    work = group->gather(output_tensors, input_tensors, opts);
+  } else {
+    int64_t group_dst_rank = getGroupRank(group, dst);
+    opts.rootRank = group_dst_rank;
+    work = group->gather(output_tensors, input_tensors, opts);
+  }
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::scatter(
+    at::Tensor tensor,
+    std::vector<at::Tensor>& scatter_list,
+    const std::shared_ptr<ProcessGroup>& group,
+    int64_t src,
+    bool async_op) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  auto my_rank = getRank(default_pg_);
+
+  std::vector<at::Tensor> output_tensors = {std::move(tensor)};
+  std::vector<std::vector<at::Tensor>> input_tensors;
```

Comment:
do same check as in DistributedC10d::gather corresponding to
https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L1796
?